### PR TITLE
Support objects on slurp

### PR DESCRIPTION
**Pull request by VidurMalik**
This PR adds support for auto-generation of nested objects and collection when reading from a hash/json

### CODE REVIEW

File: lib/schema_tools/modules/attributes.rb [hunk 1]
```diff
@@ -116,6 +118,54 @@ def schema
...
+            end
+          else
+            value
+          end
+        end
+
+        def nested_class(field_name)
+          klass = "#{base_class}::#{field_name.to_s.camelize}"
+          if is_a_defined_class?(klass)
+            klass.constantize
```
Comment (tam7t) [inline lib/schema_tools/modules/attributes.rb]: `field_name` is not user controllable is it? constantize is sometimes a path to RCE.

Comment (VidurMalik) [inline lib/schema_tools/modules/attributes.rb]: Field_name is what the user defines as the class name in the schema. It's controlled by whoever made the schema. In any case, we first check if the klass is defined, so I don't think it's a big deal here

Comment (tam7t) [inline lib/schema_tools/modules/attributes.rb]: Ah gotcah. I guess it's no different than before either :+1:

File: lib/schema_tools/modules/attributes.rb [hunk 2]
```diff
@@ -116,6 +118,54 @@ def schema
...
+          end
+        end
+
+        def base_class
+          self.to_s.deconstantize
+        end
+
+        # This is a little ugly, because we cannot use Object.const_get. That is because
+        # const_get works only if the class has already been autoloaded. Find a way to refactor
+        def is_a_defined_class?(klass)
```
Comment (tam7t) [inline lib/schema_tools/modules/attributes.rb]: Do we really care if the klass is defined? isn't it the programmers problem if they define a reference to a non-existing class?

Comment (VidurMalik) [inline lib/schema_tools/modules/attributes.rb]: That's break things like `user_data` and will enforce having a different class for each object. Not sure if we want that

File: lib/schema_tools/modules/attributes.rb [hunk 3]
```diff
@@ -76,7 +76,8 @@ def from_json(json)
         # @param [Hash{String=>Mixed}] json string or hash
         # @param [Object] obj if you want to update an existing objects
         # attributes. e.g during an update
-        def from_hash(hash, obj=nil)
+        # @param [Hash] flags for switching between ways to slurp in hash
+        def from_hash(hash, obj=nil, opts={})
```
Comment (tam7t) [inline lib/schema_tools/modules/attributes.rb]: Is there a reason opts was added?

**Comment by tam7t** [comment]
Just question about unused `opts` other than that LGTM

File: lib/schema_tools/modules/attributes.rb [hunk 4]
```diff
@@ -84,25 +85,26 @@ def from_hash(hash, obj=nil)
...
-                         end
-                       elsif field_props['type'] == 'integer'
-                         val.to_i
-                       else # rely on preceding call e.g from_json for boolean, number
-                         val
-                       end
-                      # TODO if val is a hash / array => look for nested class & cast
+            field_type = field_props['type']
+            unless val.nil?
+              case field_type
```
Comment (akatz) [inline lib/schema_tools/modules/attributes.rb:93]: This is the actual factory. I wonder if it's worth breaking it out into it's own object.

**Comment by akatz** [comment]
minor comment. LGTM.

**Pull request merged**